[automate-2629] add users for team workflow

msorens opened the pull request:
### :nut_and_bolt: Description: What code changed, and why?

Here's the user story:
We want to add users to a particular team. We drill down to the team details, select "Add Users" and, lo and behold, all users that could be added have already been added, so there are no users to add.
But now there is a new message, with a big, enticing button to press.
<img width="961" alt="image" src="https://user-images.githubusercontent.com/6817500/72999407-2efd9600-3db4-11ea-9add-3fc39224c1be.png">

The goal of this PR is for that button to open the self-same user creation modal just recently updated for the users list page. Only now, the modal opens on top of the team's "Add User" dialog (left) -- and it is a dialog in a dialog (right).
<img width="1003" alt="image" src="https://user-images.githubusercontent.com/6817500/72999644-7f74f380-3db4-11ea-9e0e-673b163fcfa3.png">

Problem: that recently finished modal was not self-contained. No fault of that development--none of our existing creation/deletion modals are. They all require intricate wiring and handlers on the base page invoking them.
In order to avoid duplicating all that identical boilerplate, I am introducing a new design pattern: **the self-contained modal**. 

* Instead of 5 or 6 or 7 inputs and outputs, it needs just one: a signal to open and the modal manages itself. 
```
<app-create-user-modal [openEvent]="openUserModal"></app-create-user-modal>
```
* Instead of several methods, variables, and event emitters to manage communication between the modal and the base page, there is just one:
```
openModal(): void {
    this.openUserModal.emit(true);
  }
```
These improve (a) encapsulation and (b) adherence to the Single Responsibility Principle (the users' list should not need to know details about adding a user).

Once I had settled on the pattern implementation was very straightforward, thanks to the new modal (thanks @bcmdarroch and @blakestier !). Most of it was simply moving code from `UserFormComponent` and `UserManagementComponent` into the new `CreateUserModalComponent`. The rest was simplifying because we no longer need to manage the larger set of inputs/outputs to the modal.

### :chains: Related Resources
PR #2654 User modal update

### :+1: Definition of Done
You can add new local users from the teams "Add Users" page when the current list has been exhausted.

### :athletic_shoe: How to Build and Test the Change

Rebuild stencil library and automate-ui server:
```
cd components/automate-ui
make update-ui-lib
make serve
```

1. Navigate to Teams >> (team) >> Add Users.
2. If there are any users available to add, add them all, then re-open Add Users.
3. Select <kbd>Create User</kbd> and exercise the dialog. Do not forgot to try error conditions (attempting to add a user with an existing username, adding a password of "eeeeeeee", etc.)
4. For a successful addition, the dialog will close and the new user will appear in the list, ready to be added to the team.
4. Navigate to Users, select <kbd>Create User</kbd>, and exercise the dialog.

### :white_check_mark: Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] Tests added/updated?
- [ ] Docs added/updated?
- [x] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

msorens commented on components/automate-ui/src/app/modules/team/team-add-users/team-add-users.component.html:
```diff
@@ -10,8 +10,10 @@
     <div id="no-users-container" *ngIf="usersNotFiltered().length === 0">
-      <p>There are no more local <a [routerLink]="['/settings', 'users']" target="_blank">users</a> to add; create some more!</p>
+      <p>There are no more local users to add; create some more!</p>
+      <chef-button primary (click)="openModal()">Create User</chef-button>
```
These are the only template changes necessary on the "Add Users" dialog for teams: the modal and the button.

msorens commented on components/automate-ui/src/app/modules/team/team-add-users/team-add-users.component.ts:
```diff
@@ -186,4 +187,9 @@ export class TeamAddUsersComponent implements OnInit, OnDestroy {
   }
+
+  openModal(): void {
+    this.openUserModal.emit(true);
```
As discussed in the preamble, this is all the wiring necessary.

msorens commented on components/automate-ui/src/app/modules/team/team-details/team-details.component.html:
```diff
@@ -78,9 +78,9 @@
-          <span *ngIf="!saving">Save</span>
+          <chef-loading-spinner *ngIf="saveInProgress"></chef-loading-spinner>
+          <span *ngIf="saveInProgress">Saving...</span>
+          <span *ngIf="!saveInProgress">Save</span>
```
PR bonus: this was broken--would never see "Save" -> "Saving..." -> "Save" on team details due to wrong variable.

msorens commented on components/automate-ui/src/app/modules/user/user-management/user-management.component.html:
```diff
@@ -9,36 +9,13 @@
         <chef-subheading>Chef Automate only displays local users.</chef-subheading>
       </chef-page-header>
-  
-      <chef-modal [visible]="createModalVisible"
```
This block of code has moved to CreateUserModalComponent

msorens commented on components/automate-ui/src/app/modules/user/user-management/user-management.component.spec.ts:
```diff
@@ -57,41 +45,10 @@ describe('UserManagementComponent', () => {
     expect(component).toBeTruthy();
   });
-
-  describe('#createUserForm', () => {
```
This set of tests has moved to CreateUserModalComponent.

msorens commented on components/automate-ui/src/app/modules/user/user-management/user-management.component.ts:
```diff
@@ -1,46 +1,28 @@
   templateUrl: './user-management.component.html',
   styleUrls: ['./user-management.component.scss']
 })
 export class UserManagementComponent implements OnInit, OnDestroy {
```
Stripping out all the modal management from the users' list. Most of this moved into CreateUserModalComponent.

msorens commented on components/automate-ui/src/app/page-components/create-user-modal/create-user-modal.component.html:
```diff
@@ -0,0 +1,92 @@
+<chef-modal [visible]="visible" (closeModal)="closeCreateModal()">
```
This template, code, scss, and tests, were essentially bulk moved from UserManagementComponent and UserFormComponent, so the behavior should be unchanged.

blakestier commented on components/automate-ui/src/app/page-components/create-user-modal/create-user-modal.component.ts:
```diff
@@ -0,0 +1,147 @@
+        this.creatingUser = false;
+        this.conflictError = false;
+        this.passwordError = false;
+        this.createUserForm.reset();
```
🎉

blakestier commented:
Looks great-- everything operating as expected for me. For some reason, opening and canceling the modal seem to lag. . . maybe it's my machine?

bcmdarroch commented on components/automate-ui/src/app/page-components/create-user-modal/create-user-modal.component.ts:
```diff
@@ -1,15 +1,140 @@
+
+  private isDestroyed = new Subject<boolean>();
+
+  ngOnChanges(changes: SimpleChanges): void {
```
we seem to be using `ngOnChanges` more lately. what changes get detected?

msorens replied:
It detects any changes on `@Input` properties. Note that whenever using `ngOnChanges` it is important to filter to the specific one you are interested in. Very much like the way we filter `@Effect` items with `ofType`, though it is a very different syntax.

bcmdarroch commented on components/automate-ui/src/app/page-components/create-user-modal/create-user-modal.component.ts:
```diff
@@ -106,6 +93,26 @@ export class CreateUserModalComponent implements OnInit, OnDestroy, OnChanges {
       takeUntil(this.isDestroyed))
       .subscribe(() => this.closeCreateModal());
+
+    combineLatest([
```
hmm so this now subscribes directly to the store instead of getting it passes from the parent component (which was user-management). I like eliminating the middle man.

bcmdarroch commented on components/automate-ui/src/app/modules/user/user-management/user-management.component.ts:
```diff
@@ -120,16 +60,7 @@ export class UserManagementComponent implements OnInit, OnDestroy {
-    this.conflictErrorEvent.emit(false);
-    this.passwordErrorEvent.emit(false);
-    this.closeEvent.emit();
+    this.openUserModal.emit(true);
```
this is a nice simplification

bcmdarroch commented on components/automate-ui/src/app/page-components/create-user-modal/create-user-modal.component.ts:
```diff
@@ -0,0 +1,147 @@
+    this.store.dispatch(new CreateUser(userCreateReq));
+  }
+
+  handleUsernameInput(event: KeyboardEvent): void {
```
actually let me know if you plan to resolve this in this PR or not - I can also fix it in the [issue](https://github.com/chef/automate/issues/1523) I'm working on

msorens replied:
Why don't you go ahead and do that in your PR. ✔️

susanev commented:
oh also, we at some point lost the padding between the label and the input, that should be 9px i believe

msorens commented:
@susanev I have made updates for all four items you requested. (Not sure why the form field widths changed but I corrected them.)
Note that one changed involved a stencil component, which changed the steps to test the PR--I have updated the instructions in the preamble above.
